src/docker/build.rs

```diff
@@ -0,0 +1,125 @@
+use std::env;
+use std::process::Command;
+use std::str::FromStr;
+
+use super::engine::Engine;
+use crate::errors::*;
+use crate::shell::Verbosity;
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum Progress {
+    Plain,
+    Auto,
+    Tty,
+}
+
+impl FromStr for Progress {
+    type Err = eyre::ErrReport;
+
+    fn from_str(progress: &str) -> Result<Self> {
+        Ok(match progress {
+            "plain" => Progress::Plain,
+            "auto" => Progress::Auto,
+            "tty" => Progress::Tty,
+            s => eyre::bail!("unexpect progress type: expected plain, auto, or tty and got {s}"),
+        })
+    }
+}
+
+impl From<Progress> for &str {
+    fn from(progress: Progress) -> Self {
+        match progress {
+            Progress::Plain => "plain",
+            Progress::Auto => "auto",
+            Progress::Tty => "tty",
+        }
+    }
+}
+
+pub trait DockerCommandExt {
+    fn build(&mut self) -> &mut Self;
+    fn progress(&mut self, progress: Option<&str>) -> Result<&mut Self>;
+    fn verbose(&mut self, verbosity: Verbosity) -> &mut Self;
+    fn scan_suggest(&mut self, suggest: bool) -> &mut Self;
+    fn cross_labels(&mut self, target: &str, platform: &str) -> &mut Self;
+}
+
+impl DockerCommandExt for Command {
+    fn build(&mut self) -> &mut Self {
+        match Engine::has_buildkit() {
+            true => self.args(["buildx", "build"]),
+            false => self.arg("build"),
+        }
+    }
+
+    fn progress(&mut self, progress: Option<&str>) -> Result<&mut Self> {
+        let progress: Progress = match progress {
+            None => env::var("CROSS_BUILD_PROGRESS")
+                .as_deref()
+                .unwrap_or("auto")
+                .parse(),
+            Some(progress) => progress.parse(),
+        }?;
+        Ok(self.args(["--progress", progress.into()]))
+    }
+
+    fn verbose(&mut self, verbosity: Verbosity) -> &mut Self {
+        match verbosity {
+            Verbosity::Verbose(2..) => self.args(["--build-arg", "VERBOSE=1"]),
+            _ => self,
+        }
+    }
+
+    fn scan_suggest(&mut self, suggest: bool) -> &mut Self {
+        let value = match suggest {
+            true => "true",
+            false => "false",
+        };
+        self.env("DOCKER_SCAN_SUGGEST", value)
+    }
+
+    fn cross_labels(&mut self, target: &str, platform: &str) -> &mut Self {
+        self.args([
+            "--label",
+            &format!("{}.for-cross-target={target}", crate::CROSS_LABEL_DOMAIN,),
+        ]);
+        self.args([
+            "--label",
+            &format!("{}.runs-with={platform}", crate::CROSS_LABEL_DOMAIN,),
+        ])
+    }
+}
+
+pub trait DockerResultExt {
+    fn engine_warning(self, engine: &Engine) -> Result<()>;
+    fn buildkit_warning(self) -> Result<()>;
+}
+
+impl DockerResultExt for Result<()> {
+    fn engine_warning(self, engine: &Engine) -> Result<()> {
```

Comment:
Allow the BuildKit and engine warnings when doing local image builds as well, and not just custom ones. I'm not sure if this would interfere with GHA?

Reply:
should be fine

Reply:
only important to make sure we don't read `CI` or other ci envs when doing custom builds unless the intent is to actually do that, like https://github.com/cross-rs/cross/issues/659